src/transformers/models/phi/modeling_phi.py

```diff
@@ -0,0 +1,1339 @@
+
+        self.dim = dim
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
+        self.register_buffer("inv_freq", inv_freq)
+
+        # Build here to make `torch.jit.trace` work.
+        self._set_cos_sin_cache(
+            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
+        )
+
+    def _set_cos_sin_cache(self, seq_len, device, dtype):
+        self.max_seq_len_cached = seq_len
+        t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)
+
+        freqs = torch.einsum("i,j->ij", t, self.inv_freq)
+        emb = freqs
+        self.register_buffer("cos_cached", emb.cos().to(dtype), persistent=False)
+        self.register_buffer("sin_cached", emb.sin().to(dtype), persistent=False)
+
+    def forward(self, x, seq_len=None):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if seq_len > self.max_seq_len_cached:
+            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
+
+        return (
+            self.cos_cached[:seq_len, ...].to(dtype=x.dtype),
+            self.sin_cached[:seq_len, ...].to(dtype=x.dtype),
+        )
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids, rotary_dim):
+    q = q.transpose(1, 2)
+    k = k.transpose(1, 2)
+    cos = cos[position_ids].unsqueeze(2)
+    sin = sin[position_ids].unsqueeze(2)
+
+    # seperate which portions to apply rotary embeddings to and which ones to pass.
+    q_rot = q[:, :, :, :rotary_dim]
+    q_pass = q[:, :, :, rotary_dim:]
+    k_rot = k[:, :, :, :rotary_dim]
+    k_pass = k[:, :, :, rotary_dim:]
+
+    # Splits the queries and keys in half
+    q1, q2 = q_rot.chunk(2, dim=-1)
+    k1, k2 = k_rot.chunk(2, dim=-1)
+
+    # apply sin and cos
+    q_rot = torch.cat([q1 * cos - q2 * sin, q1 * sin + q2 * cos], axis=-1)
+    k_rot = torch.cat([k1 * cos - k2 * sin, k1 * sin + k2 * cos], axis=-1)
+
+    # concatenate the `_pass` and `_rot` embeddings.
+    q_embed = torch.cat([q_rot, q_pass], dim=-1).transpose(1, 2)
+    k_embed = torch.cat([k_rot, k_pass], dim=-1).transpose(1, 2)
+
+    return q_embed, k_embed
+
+
+# Copied from transformers.models.clip.modeling_clip.CLIPMLP with CLIP->Phi
+class PhiMLP(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.activation_fn = ACT2FN[config.hidden_act]
+        self.fc1 = nn.Linear(config.hidden_size, config.intermediate_size)
+        self.fc2 = nn.Linear(config.intermediate_size, config.hidden_size)
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        hidden_states = self.fc1(hidden_states)
+        hidden_states = self.activation_fn(hidden_states)
+        hidden_states = self.fc2(hidden_states)
+        return hidden_states
+
+
+class PhiAttention(nn.Module):
```

Comment:
You should have a look at `PersimmonAtttention` ! It also has a rotary embedding applied to only a portion of the input.

Reply:
The attention class for `Phi` is almost same for `Persimmon` except the slight difference in `_split_heads` method. 
I have added `# Copied from` statements to all methods(init, forward, _init_rope) in the attention class except the _split_heads.